File: samples/java-spring-rest-assured/src/main/java/art/galushko/java/spring/rest/assured/config/ApiKeyFilter.java
package art.galushko.java.spring.rest.assured.config;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

@Component
public class ApiKeyFilter extends OncePerRequestFilter {

    private final String expectedApiKey;

    public ApiKeyFilter(@Value("${app.security.apiKey}") String expectedApiKey) {
        this.expectedApiKey = expectedApiKey;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
            throws ServletException, IOException {
        String apiKey = request.getHeader("X-API-Key");
        if (expectedApiKey != null && !expectedApiKey.isBlank()) {
            if (!expectedApiKey.equals(apiKey)) {
                response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
                response.setContentType("application/json");
                response.getWriter().write("{\"code\":\"unauthorized\",\"message\":\"API key required\"}");
                return;
            }
        }
        filterChain.doFilter(request, response);
    }
}

File: samples/kotlin-spring-rest-assured/src/main/kotlin/art/galushko/kotlin/spring/rest/assured/config/ApiKeyFilter.kt
package art.galushko.kotlin.spring.rest.assured.config

import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import java.io.IOException

@Component
class ApiKeyFilter(
    @param:Value("\${app.security.apiKey}")
    private val expectedApiKey: String
) : OncePerRequestFilter() {

    @Throws(ServletException::class, IOException::class)
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain
    ) {
        val apiKey = request.getHeader("X-API-Key")
        if (expectedApiKey.isNotBlank()) {
            if (expectedApiKey != apiKey) {
                response.status = HttpServletResponse.SC_UNAUTHORIZED
                response.contentType = "application/json"
                response.writer.write("{\"code\":\"unauthorized\",\"message\":\"API key required\"}")
                return
            }
        }
        filterChain.doFilter(request, response)
    }
}
